File: CPP/Algorithm/Generate_Parentheses.cpp
// Complexity: O(2^n)

void gen(auto &s, auto open, auto close, auto &v) {
    if (!open and !close) {
        v.emplace_back(s); return;
    }
    if (open) {
        s += '('; gen(s, open - 1, close, v);
        s.pop_back();  // Back_tracking
    }
    if (close) {
        if (open < close) {
            s += ')'; gen(s, open, close - 1, v);
            s.pop_back();  // Back_tracking
        }
    }
}

// Calling: gen(empty_str, n, n, empty_vec)

File: Python/Algorithm/Generate_Parentheses.py
# Complexity: O(2^n)

def gen(s, open, close, com):
    if open == 0 and close == 0: com.append(s); return
    if open > 0: gen(s + '(', open - 1, close, com)
    if close > 0 and open < close: gen(s + ')', open, close - 1, com)

# Example of calling the function
empty_str = ""
n = 3  # Example with 3 pairs of parentheses
empty_vec = []
gen(empty_str, n, n, empty_vec)
for combination in empty_vec:
    print(combination)